File: iosApp/iosApp/feature/profile/order/order_details/OrderDetailsViewModel.swift
//
//  OrderDetailsViewModel.swift
//  iosApp
//
//  Created by Марк Шавловский on 12.07.2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import shared

class OrderDetailsViewModel :ObservableObject {

    @Published var orderDetailsViewState:OrderDetailsViewState = OrderDetailsViewState(
        code: "",
        status: OrderStatus.notAccepted,
        statusName: "",
        dateTime: "",
        pickupMethod: "",
        deferredTimeHintString: "",
        deferredTime: nil,
        address: "",
        comment: nil,
        deliveryCost: nil,
        isDelivery: false,
        oldAmountToPay: nil,
        newAmountToPay: "",
        orderProductList: []
    )


    let dateUtil = DateUtil()
    
    init(uuid:String){
        iosComponent.provideIOrderInteractor().observeOrderByUuid(orderUuid: uuid).watch { orderWithAmout in
            (self.orderDetailsViewState.copy() as! OrderDetailsViewState).apply { copiedState in
                copiedState.code = orderWithAmout?.code ?? ""
                copiedState.status = orderWithAmout?.status ?? OrderStatus.accepted
                copiedState.statusName = OrderChip.getStatusName(status: (orderWithAmout?.status ?? OrderStatus.accepted))
                
                if(orderWithAmout?.dateTime != nil){
                    copiedState.dateTime = self.dateUtil.getDateTimeString(dateTime: orderWithAmout!.dateTime)
                }
                
                if(orderWithAmout?.isDelivery == true){
                    copiedState.pickupMethod = "Доставка"
                    copiedState.deferredTimeHintString = "Ко времени"

                }else{
                    copiedState.pickupMethod = "Самовывоз"
                    copiedState.deferredTimeHintString = "Ко времени"
                }
                
                if(orderWithAmout?.deferredTime != nil){
                    copiedState.deferredTime = self.dateUtil.getTimeString(time: orderWithAmout!.deferredTime!)
                }

                copiedState.address = orderWithAmout?.address ?? ""
                copiedState.comment = orderWithAmout?.comment ?? ""

                copiedState.isDelivery = orderWithAmout?.isDelivery ?? false
                if(orderWithAmout?.oldAmountToPay != nil){
                    copiedState.oldAmountToPay =  "\(orderWithAmout?.oldAmountToPay ?? 0)\(Strings.CURRENCY)"
                }
                copiedState.newAmountToPay = "\(orderWithAmout?.newAmountToPay ?? 0)\(Strings.CURRENCY)"

                copiedState.orderProductList = orderWithAmout?.orderProductList.map({ orderProductsWithCosts in
                    var oldPrice:String? = nil
                    var oldCost:String? = nil
                    
                    if(orderProductsWithCosts.product.oldPrice != nil){
                        oldPrice = "\(orderProductsWithCosts.product.oldPrice ?? 0)\(Strings.CURRENCY)"
                    }
                    
                    if(orderProductsWithCosts.oldCost != nil){
                        oldCost = "\(orderProductsWithCosts.oldCost ?? 0)\(Strings.CURRENCY)"
                    }
                    
                    return OrderProductItem(id: orderProductsWithCosts.uuid,
                                     name: orderProductsWithCosts.product.name,
                                     newPrice: "\(orderProductsWithCosts.product.newPrice)\(Strings.CURRENCY)",
                                     oldPrice: oldPrice,
                                     newCost: "\(orderProductsWithCosts.newCost)\(Strings.CURRENCY)",
                                     oldCost: oldCost,
                                     photoLink: orderProductsWithCosts.product.photoLink,
                                     count: "× \(orderProductsWithCosts.count)")
                }) ?? []
                if(orderWithAmout?.deliveryCost != nil){
                    copiedState.deliveryCost = "\(orderWithAmout?.deliveryCost ?? 0)\(Strings.CURRENCY)"
                }
                self.orderDetailsViewState = copiedState
            }
        }
    }

}

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/order_details/OrderDetailsViewModel.kt
package com.bunbeauty.shared.presentation.order_details

import com.bunbeauty.shared.Constants.PERCENT
import com.bunbeauty.shared.domain.feature.order.ObserveOrderUseCase
import com.bunbeauty.shared.domain.feature.order.StopObserveOrdersUseCase
import com.bunbeauty.shared.domain.model.order.Order
import com.bunbeauty.shared.extension.launchSafe
import com.bunbeauty.shared.presentation.base.SharedStateViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collectLatest

class OrderDetailsViewModel(
    private val observeOrderUseCase: ObserveOrderUseCase,
    private val stopObserveOrdersUseCase: StopObserveOrdersUseCase,
) : SharedStateViewModel<OrderDetails.ViewDataState, OrderDetails.Action, OrderDetails.Event>(
    OrderDetails.ViewDataState(
        orderDetailsData = OrderDetails.ViewDataState.OrderDetailsData(
            orderProductItemList = listOf(),
            orderInfo = null,
            oldTotalCost = null,
            deliveryCost = null,
            newTotalCost = null,
            discount = null
        ),
        screenState = OrderDetails.ViewDataState.ScreenState.LOADING,
    )
) {

    private var observeOrderJob: Job? = null
    private var orderObservationUuid: String? = null

    override fun reduce(action: OrderDetails.Action, dataState: OrderDetails.ViewDataState) {
        when (action) {
            OrderDetails.Action.Back -> addEvent {
                OrderDetails.Event.Back
            }

            is OrderDetails.Action.Init -> loadOrder(action.orderUuid)
            OrderDetails.Action.StopObserve -> stopObserveOrders()
        }
    }

    private fun loadOrder(orderUuid: String) {
        setState {
            copy(
                screenState = OrderDetails.ViewDataState.ScreenState.LOADING,
            )
        }

        observeOrderJob = sharedScope.launchSafe(
            block = {
                val (uuid, orderFlow) = observeOrderUseCase(orderUuid)
                orderObservationUuid = uuid
                orderFlow.collectLatest { order ->
                    if (order != null) {
                        setState {
                            copy(
                                orderDetailsData = OrderDetails.ViewDataState.OrderDetailsData(
                                    orderProductItemList = getProductList(order),
                                    orderInfo = getOrderInfo(order),
                                    deliveryCost = order.deliveryCost?.toString(),
                                    oldTotalCost = order.oldTotalCost?.toString(),
                                    newTotalCost = order.newTotalCost.toString(),
                                    discount = order.percentDiscount?.let { discount ->
                                        discount.toString() + PERCENT
                                    }
                                ),
                                screenState = OrderDetails.ViewDataState.ScreenState.SUCCESS,
                            )
                        }
                    }
                }
            },
            onError = {

            },
        )

    }

    private fun stopObserveOrders() {
        observeOrderJob?.cancel()
        orderObservationUuid?.let { uuid ->
            sharedScope.launchSafe(
                block = {
                    stopObserveOrdersUseCase(uuid)
                },
                onError = {

                },
            )
        }
        orderObservationUuid = null
    }

    private fun getOrderInfo(order: Order) =
        OrderDetails.ViewDataState.OrderDetailsData.OrderInfo(
            code = order.code,
            status = order.status,
            dateTime = order.dateTime,
            deferredTime = order.deferredTime,
            address = order.address,
            comment = order.comment,
            isDelivery = order.isDelivery,
            paymentMethod = order.paymentMethod
        )

    private fun getProductList(order: Order) =
        order.orderProductList.map { orderProduct ->
            OrderDetails.ViewDataState.OrderDetailsData.OrderProductItem(
                uuid = orderProduct.uuid,
                name = orderProduct.product.name,
                newPrice = orderProduct.product.newPrice.toString(),
                oldPrice = orderProduct.product.oldPrice?.toString(),
                newCost = (orderProduct.product.newPrice * orderProduct.count).toString(),
                oldCost = orderProduct.product.oldPrice?.let { oldPrice ->
                    (oldPrice * orderProduct.count).toString()
                },
                photoLink = orderProduct.product.photoLink,
                count = orderProduct.count.toString(),
                additions = orderProduct.orderAdditionList
            )
        }

}
